music_assistant/server/providers/tidal/__init__.py

```diff
@@ -0,0 +1,382 @@
+        result = []
+        tracks = await get_artist_toptracks(self._tidal_session, prov_artist_id)
+        for index, track_obj in enumerate(tracks, 1):
+            if track_obj.available:
+                track = parse_track(tidal_provider=self, track_obj=track_obj)
+                track.position = index
+                result.append(track)
+        return result
+
+    async def get_playlist_tracks(self, prov_playlist_id) -> AsyncGenerator[Track, None]:
+        """Get all playlist tracks for given playlist id."""
+        tracks = await get_playlist_tracks(self._tidal_session, prov_playlist_id=prov_playlist_id)
+        for index, track_obj in enumerate(tracks):
+            if track_obj.available:
+                track = parse_track(tidal_provider=self, track_obj=track_obj)
+                track.position = index + 1
+                yield track
+
+    async def get_similar_tracks(self, prov_track_id, limit=25) -> list[Track]:
+        """Get similar tracks for given track id."""
+        similar_tracks_obj = await get_similar_tracks(self._tidal_session, prov_track_id, limit)
+        tracks = []
+        for track_obj in similar_tracks_obj:
+            if track_obj.available:
+                track = parse_track(tidal_provider=self, track_obj=track_obj)
+                tracks.append(track)
+        return tracks
+
+    async def library_add(self, prov_item_id, media_type: MediaType):
+        """Add item to library."""
+        return await library_items_add_remove(
+            self._tidal_session,
+            self._tidal_user_id,
+            prov_item_id,
+            media_type,
+            add=True,
+        )
+
+    async def library_remove(self, prov_item_id, media_type: MediaType):
+        """Remove item from library."""
+        return await library_items_add_remove(
+            self._tidal_session,
+            self._tidal_user_id,
+            prov_item_id,
+            media_type,
+            add=False,
+        )
+
+    async def add_playlist_tracks(self, prov_playlist_id: str, prov_track_ids: list[str]):
+        """Add track(s) to playlist."""
+        return await add_remove_playlist_tracks(
+            self._tidal_session, prov_playlist_id, prov_track_ids, add=True
+        )
+
+    async def remove_playlist_tracks(
+        self, prov_playlist_id: str, positions_to_remove: tuple[int, ...]
+    ) -> None:
+        """Remove track(s) from playlist."""
+        prov_track_ids = []
+        async for track in self.get_playlist_tracks(prov_playlist_id):
+            if track.position in positions_to_remove:
+                prov_track_ids.append(track.item_id)
+            if len(prov_track_ids) == len(positions_to_remove):
+                break
+        return await add_remove_playlist_tracks(
+            self._tidal_session, prov_playlist_id, prov_track_ids, add=False
+        )
+
+    async def create_playlist(self, name: str) -> Playlist:  # type: ignore[return]
```

Comment:
why the type ignore here ?I think that can be omitted

Reply:
Removed, and yes, remnant of copy paste from other provider